File: Python/ByRound/2059/2059_D_Graph_and_Graph.py
# Problem: CF 2059 D - Graph and Graph
# https://codeforces.com/contest/2059/problem/D

"""
D. Graph and Graph

Purpose:
This problem involves finding the minimum total cost of moving two tokens simultaneously in two different graphs, where each token moves to adjacent vertices in its respective graph at each step. The cost of a move is defined as the absolute difference between the new positions of the two tokens.

Approach:
- Use a modified Dijkstra-like algorithm with a deque (0-1 BFS) to find the minimum cost path.
- Key idea: the state is represented by (v1, v2), where v1 and v2 are current positions of tokens in graph 1 and 2 respectively.
- Stop vertices are those where the two graphs share an edge between the same vertex (intersection of neighbors).
- For every valid move from a token at position u1 in graph 1 to u2 in graph 2, we calculate cost = offset + |u1 - u2|.
- We use a deque of queues based on cost for 0-1 BFS, allowing for better efficiency compared to regular Dijkstra.

Time Complexity: O(n * (m1 + m2)) per test case.
Space Complexity: O(n^2 + m1 + m2) for storing adjacency lists and seen matrix.

Algorithms/Techniques:
- 0-1 BFS using deque with buckets
- Graph traversal with state tracking
"""

import sys
from collections import deque


def input_general():
    return sys.stdin.readline().rstrip("\r\n")


def input_num():
    return int(sys.stdin.readline().rstrip("\r\n"))


def input_multi(x=int):
    return map(x, sys.stdin.readline().rstrip("\r\n").split())


def input_list(x=int):
    return list(input_multi(x))


def main():
    n, s1, s2 = input_multi()
    m1 = input_num()
    ed1 = [[] for _ in range(n + 1)]
    for _ in range(m1):
        u, v = input_multi()
        ed1[u].append(v)
        ed1[v].append(u)
    m2 = input_num()
    ed2 = [[] for _ in range(n + 1)]
    for _ in range(m2):
        u, v = input_multi()
        ed2[u].append(v)
        ed2[v].append(u)
    stop = set()
    # Identify vertices where both graphs have a common neighbor
    for u in range(1, n + 1):
        if set(ed1[u]).intersection(set(ed2[u])):
            stop.add(u)
    seen = [[10**10 for _ in range(n + 1)] for _ in range(n + 1)]
    seen[s1][s2] = 0
    dq = deque([[(s1, s2)]])  # Initialize queue with starting state (s1,s2)
    offset = 0
    in_stack = 1  # Counter of elements to be processed
    while in_stack:
        # Remove empty queues
        while len(dq[0]) == 0:
            offset += 1
            dq.popleft()
        v1, v2 = dq[0].pop()  # Get next state to process
        in_stack -= 1
        if v1 == v2 and v1 in stop:  # If both tokens are at same position and it's a stop vertex
            print(offset)
            return
        if offset > seen[v1][v2]:  # Skip outdated states
            continue
        for u1 in ed1[v1]:  # Move token in first graph
            for u2 in ed2[v2]:  # Move token in second graph
                new = abs(u1 - u2)  # Cost to reach new state
                cost = offset + new
                if cost < seen[u1][u2]:  # Update minimum cost if better path found
                    seen[u1][u2] = cost
                    while len(dq) <= new:  # Extend deque if needed
                        dq.append([])
                    dq[new].append((u1, u2))  # Add new state to appropriate queue
                    in_stack += 1  # Increment counter
    print(-1)  # No solution found


if __name__ == "__main__":
    cases = input_num()
    for _ in range(cases):
        main()



# https://github.com/VaHiX/codeForces/

File: CPP/ByTier/D/2059_D_Graph_and_Graph.cpp
// Problem: CF 2059 D - Graph and Graph
// https://codeforces.com/contest/2059/problem/D

/*
D. Graph and Graph
time limit per test2 seconds
memory limit per test256 megabytes

Algorithm: Dijkstra's shortest path on a state space (u, v) where u is node in first graph and v is node in second graph.
We compute the minimum cost to reach a state where both tokens are at same vertex that is also a common vertex (i.e., exists in both graphs).

Time Complexity: O((V + E1 + E2) * log(V^2)) = O((V + E1 + E2) * log V)
Space Complexity: O(V^2 + E1 + E2)

The problem models a two-token game on two graphs, where at each step,
the tokens move to adjacent nodes in their respective graphs and the cost is |u1 - u2|.
We want to find the minimum total cost to reach a state where both tokens are at the same vertex
that exists in both graphs (i.e., has common neighbors in both).
This is modeled as a weighted directed graph with states defined by pairs of nodes from each graph.
*/
#include <stdlib.h>
#include <algorithm>
#include <functional>
#include <iostream>
#include <queue>
#include <unordered_set>
#include <vector>

using namespace std;
#define FOR(i, n) for (int i = 0; i < (n); i++)
#define FOR1(i, n) for (int i = 1; i <= (n); i++)
#define ll long long
struct Node {
  int u, v, cost;
  bool operator>(const Node &other) const { return cost > other.cost; }
};
void solve() {
  int N, st1, st2;
  cin >> N >> st1 >> st2;
  int E1;
  cin >> E1;
  vector<vector<int>> g1(N + 1), g2(N + 1);
  FOR(i, E1) {
    int x, y;
    cin >> x >> y;
    g1[x].push_back(y);
    g1[y].push_back(x);
  }
  int E2;
  cin >> E2;
  FOR(i, E2) {
    int x, y;
    cin >> x >> y;
    g2[x].push_back(y);
    g2[y].push_back(x);
  }
  vector<bool> com(N + 1, false); // Mark nodes common to both graphs
  FOR1(i, N) {
    unordered_set<int> s;
    for (int nb : g1[i])
      s.insert(nb);
    for (int nb : g2[i]) {
      if (s.count(nb)) {  // If neighbor exists in both graphs
        com[i] = true;
        break;
      }
    }
  }
  const int INF = 1000000000;
  vector<vector<int>> d(N + 1, vector<int>(N + 1, INF)); // d[a][b] = min cost to reach state (a,b)
  priority_queue<Node, vector<Node>, greater<Node>> pq;  // Min-heap for Dijkstra
  d[st1][st2] = 0;
  pq.push({st1, st2, 0});
  int ans = -1;
  while (!pq.empty()) {
    Node cur = pq.top();
    pq.pop();
    int a = cur.u, b = cur.v, curd = cur.cost;
    if (curd != d[a][b]) // Skip outdated entries
      continue;
    if (a == b && com[a]) {  // If both tokens are at same node which is common in both
      ans = curd;
      break;
    }
    for (int nxt1 : g1[a])
      for (int nxt2 : g2[b]) {
        int nd = curd + abs(nxt1 - nxt2);
        if (nd < d[nxt1][nxt2]) {  // Relax edge and push to heap
          d[nxt1][nxt2] = nd;
          pq.push({nxt1, nxt2, nd});
        }
      }
  }
  cout << ans << "\n";
}
int main() {
  ios::sync_with_stdio(false);
  cin.tie(nullptr);
  cout.tie(nullptr);
  int T;
  cin >> T;
  while (T--) {
    solve();
  }
  return 0;
}


// https://github.com/VaHiX/codeForces/